### PR TITLE
Support dynamic batch size in online_norm for tensorflow

### PR DESCRIPTION
**Pull request by Ajk4**
Currently in online_norm (not batch_online_norm) we store intermediate
outputs from forward pass for backward pass in a tensorflow variable. Tensorflow variables have fixed shape which makes it impossible to use one OnlineNorm layer in two places with different batch sizes.

In my use case (and I assume generally in Reinforcement Learning) I am using one tensorflow layers
with different batch sizes in two places - in policy rollouts and in policy updates. 

I noticed that `outputs` can be passed from forward to backward via reference (and not via mutable tensorflow variable).

In this PR I implement that change making OnlineNorm layer batch-size agnostic and getting rid of unnecessary b_dim argument.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+.idea
+online-norm/tensorflow/online_norm_tf.egg-info
+__pycache__
```
Comment (vchiley) [inline .gitignore:3]: new line

**Comment by vchiley** [comment]
ON is created to run at batch size one. If the batch size is larger than batch size one then you would need to loop through the samples in the batch. If you know the batch size apriori, you can play some tricks to get around having to loop through the batch. This is implemented in [batch_online_norm.py ](https://github.com/Ajk4/online-normalization/blob/support_dynamic_batch_size/online-norm/tensorflow/online_norm_tf/batch_online_norm.py)

While this code does run it will not be doing what the ON paper advertises.

**Comment by vchiley** [comment]
but passing `output` from forward to backward via reference is a good idea.

**Comment by Ajk4** [comment]
Thanks for reply! 

So batch_size in 'online_norm.py ' must be equal to 1? 
And for B>1 I should use batch_online_norm.py?

It the 'batch_size=1' parameter kept for documentation purposes? Or is there some use-case for batch_size > 1 in 'online_norm.py' ?

**Comment by vchiley** [comment]
```
So batch_size in 'online_norm.py ' must be equal to 1?
And for B>1 I should use batch_online_norm.py?
```
is correct.
By definition Online Normalization operates on one sample at a time. `online_norm` can be used on a batch of data if you put it into a loop which processes one sample at a time. This loop is then slow, so we vectorized the operation in `batch_online_norm` using a FIR filter to replace the loop. This FIR filter approach is much faster but needs to know the batch size apriori. Another alternative would be to write a CUDA kernel for online norm.

In the `online_norm` code `batch_size=1` was for different norm technique we were looking into. It was similar but different. I think I left it there on accident when I was cleaning up the code for publishing and could probably be removed (especially if we use a reference for `output`).


Also all of my experiments were done in PyTorch.

**Comment by vchiley** [comment]
Support for a dynamic batch sizes requires a loop to loop over the samples in the batch. In TF / PyTorch / C this is slow. The best solution is to write a CUDA kernel which does the looping. I have done a first pass at a CUDA implementation and while it is fast our organization has decided not to pursue perfecting the kernel for publishing.

**Comment by vchiley** [comment]
Updt:
I wrote a CUDA kernel that accelerates OnlineNorm. It has been incorporated into PyTorch and will be integrated into the TensorFlow implementation shortly. I'll update you again when this is done.
Once integrated, this should enable dynamic batch sizes.
-Vitaliy

**Comment by vchiley** [comment]
Updt:
With https://github.com/Cerebras/online-normalization/pull/12 dynamic batch sizes are fully supported.
